packages/ethereum/src/events/index.ts

```diff
@@ -20,85 +20,234 @@ export default class Event {
+        const withdraws: WithdrawEvent[] = [];
+        const refunds: RefundEvent[] = [];
 
-        return result;
-    }
+        const logsFilter = {
+            fromBlock: fromBlock || this.config.originBlock, // TODO: constants
+            toBlock: toBlock || 'latest',
+            address: this.config.contractAddress,
+        };
 
-    async subscribe(onMessage: Function, filter?: Function) {
-        this.contract.contract.on('NewContract', (...args: []) => {
-            const swap = TransformNewContract(args);
-            const result = filter ? filter(swap) : swap;
-            if (result) {
-                onMessage(result);
+        const logs = await this.contract.provider.getLogs(logsFilter);
+
+        logs.forEach((log: Log) => {
+            const parsed = this.interface.parseLog(log);
+
+            switch (parsed.name) {
+                case 'NewContract': {
+                    const swap = {
+                        network: 'ETH',
+                        eventName: 'NEW_CONTRACT',
+                        id: parsed.values.id,
+                        hashLock: parsed.values.hashLock,
+                        sender: parsed.values.sender,
+                        receiver: parsed.values.receiver,
+                        inputAmount: parsed.values.inputAmount.toString(),
+                        outputAmount: parsed.values.outputAmount.toString(),
+                        expiration: parsed.values.expiration.toString(),
+                        outputNetwork: parsed.values.outputNetwork,
+                        outputAddress: parsed.values.outputAddress,
+                        transactionHash: log.transactionHash,
+                    };
+
+                    if (_filter.new) {
+                        if (filter(_filter.new, swap)) {
+                            swaps.push(swap);
+                        }
+                    } else {
+                        swaps.push(swap);
+                    }
+
+                    break;
+                }
+
+                case 'Withdraw': {
+                    const withdraw = {
+                        network: 'ETH',
+                        eventName: 'WITHDRAW',
+                        id: parsed.values.id,
+                        secret: parsed.values.secret,
+                        hashLock: parsed.values.hashLock,
+                        sender: parsed.values.sender,
+                        receiver: parsed.values.receiver,
+                        transactionHash: log.transactionHash,
+                    };
+
+                    if (_filter.withdraw) {
+                        if (filter(_filter.withdraw, withdraw)) {
+                            withdraws.push(withdraw);
+                        }
+                    } else {
+                        withdraws.push(withdraw);
+                    }
+
+                    break;
+                }
+
+                case 'Refund': {
+                    const refund = {
```

Comment:
case Withdraw and Refund are pretty much the same. Better approach might be to create function which returns object. The function will look like this:
const objectState = (eventName, parsed, log) => ({network: 'ETH', eventName, ...parsed.values, transactionHash: log.transactionHash})
And then
const refund/withdraw = {...objectState(eventName, parsed, log)}.
Reusable on lines: 116, 139, 206, 229 (where we are creating refund/withdraw objects)